Add youtube music

R00BEL opened the pull request.

R00BEL committed:
fix: changing the database schema

R00BEL committed:
wip: add YoutubeMusicStore

R00BEL committed:
wip: add initializeClient in YoutubeMusicService

R00BEL committed:
wip: transition to version 18 of the node
wip: yandex-short-api -> ym-api-meowed

R00BEL committed:
wip: add getPlaylistTracks and searchTrackByName in YoutubeMusicService

R00BEL committed:
wip: add addTracksToPlaylist and removeTracksFromPlaylist in YoutubeMusicService

R00BEL committed:
wip: add youtubeMusicService in SyncService

R00BEL committed:
fix: fixed a bug with deleting elements from playlists

R00BEL committed:
wip: removing restrictions when receiving an item from a playlist

R00BEL committed:
wip: searchTrackByName in YoutubeMusicService selects only tracks that match the track name and author name

R00BEL committed:
fix: when libmuse crashes, the function is re-run

R00BEL committed:
added the ability to separate the song title into title and producer name

R00BEL committed:
refactoring: YoutubeMusic described in more detail

R00BEL committed:
wip: added language parameter to YoutubeMusicService

R00BEL committed:
tracks are recorded in cache when they are in YoutubeMusicService

R00BEL committed:
wip: added a switch "isRemoveTracksThatAreNotInOriginalPlaylistAndAreNoLongerAvailable" to the config

R00BEL committed:
wip: improved ArtistToNameArray feature

R00BEL committed:
upgrading node version to 21 // fixes ts-node to work with node 21

R00BEL committed:
wip: ym-api-meowed -> yandex-short-api

R00BEL committed:
wip: added getting artists from decomposed properties when getting them in getPlaylistTracks in YoutubeMusicService

R00BEL committed:
wip: add isNoRemoveTracksThatAreNotInOriginalPlaylistAndAreNoLongerAvailable in confing

R00BEL committed:
wip: add language in YandexMusicService

R00BEL committed:
wip: add language in YandexMusicService

R00BEL committed:
fix(youtube-music): handle tracks without videoId during playlist synchronization

R00BEL committed:
refactoring: minor changes

R00BEL committed:
style(entities.ts): specify file extensions in export statements

pkarpovich commented on package.json:
```diff
@@ -65,10 +68,12 @@
     "signale": "^1.4.0",
     "spotify-web-api-node": "^5.0.2",
     "yandex-short-api": "^1.0.0",
+    "ym-api-meowed": "^1.3.1",
```
Remove unused dependency

pkarpovich commented on src/config/sync-config.ts:
```diff
@@ -3,14 +3,16 @@ import { object, array, string } from 'yup';
 
+interface TargetPlaylists
+    extends Omit<PlaylistConfig, 'excludedTrackIds' | 'targetPlaylists'> {
+    isNoRemoveTracksThatAreNotInOriginalPlaylistAndAreNoLongerAvailable?: boolean;
```
The variable name `isNoRemoveTracksThatAreNotInOriginalPlaylistAndAreNoLongerAvailable` is overly verbose and confusing. Consider renaming it to something succinct and intuitive like `unavailableTracks`. This will improve code readability and maintainability.

pkarpovich commented on src/container.ts:
```diff
@@ -17,25 +17,32 @@ import {
     SpotifyController,
     HealthController,
 } from './controllers.js';
+import { YoutubeMusicService } from './services/music-providers/youtube-music/youtube-music.service.js';
```
move to services import

pkarpovich commented on src/services/music-providers/spotify.service.ts:
```diff
@@ -68,7 +70,11 @@ export class SpotifyService implements BaseMusicService {
         this.isReady = true;
 
-        await this.authStore.set({ refreshToken: body.refresh_token });
+        const store = this.store.get();
```
Consider refactoring this code by moving the get﻿operation inside theset﻿ method. It will lead to cleaner, more maintainable code by reducing redundancy and the chance of state inconsistency

R00BEL committed:
fix: removed unused package ym-api-meowed
refactoring: isNoRemoveTracksThatAreNotInOriginalPlaylistAndAreNoLongerAvailable -> hasUnavailableTracks

pkarpovich commented on src/services/music-providers/yandex-music.service.ts:
```diff
@@ -54,4 +65,16 @@ export class YandexMusicService extends BaseMusicService {
+            return [artist.name];
+        }
+
+        const artistsInDecomposed = artist.decomposed.filter(
```
Please consider renaming `artistsInDecomposed` to a more indicative name like `filteredArtists`. It enhances code readability, making it easier to understand the purpose of the variable.

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+
+export class YoutubeMusicService extends BaseMusicService {
+    isReady = false;
+    private getPlaylistTracksLimit = 99999999999;
```
The current `getPlaylistTracksLimit` value is set to an arbitrary large number, which is not a recommended practice. Instead, use JavaScript's `Number.MAX_VALUE` to represent the largest permissible number. This makes the code more readable, maintainable, and prevents possible overflow errors.

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+        );
+    }
+
+    private notifyThatFunctionCompletedWithError(functionName: string) {
```
Consider renaming the method `notifyThatFunctionCompletedWithError` to something more concise and meaningful, such as `logErrorAndRetry`. This will improve readability and better reflect the method's purpose: logging an error and initiating a retry.

pkarpovich replied:
Please consider refactoring ﻿notifyThatFunctionCompletedWithError to use ﻿arguments.callee.name instead of manually passing the function name. This will make the code DRY, more maintainable, and less prone to human errors.

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+        return result;
+    }
+
+    private checkThatTrackIsSuitable(
```
Consider renaming `checkThatTrackIsSuitable` to a more descriptive name like `isTrackMatchingCriteria`. This improves readability and makes it immediately clear what the function is checking.

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+        return checkIfArraysAreEqual<string>(findArtist, allFoundArtists);
+    }
+
+    private getTrackNameWithoutParticipantDesignations(name: string): string {
```
The function name `getTrackNameWithoutParticipantDesignations` seems to be verbose and lacks precision. Consider renaming it to `extractTrackName` or `getPrimaryTrackName` for clarity and simplicity, improving readability and maintainability of the code.

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+        foundName = foundName.toLowerCase();
+        foundArtists = foundArtists.map((a) => a.toLowerCase());
+
+        const findNameWithoutParticipantDesignations =
```
Consider renaming `findNameWithoutParticipantDesignations` and `foundNameWithoutParticipantDesignations` to be more concise and descriptive. For example, `processedFindName` and `processedFoundName` could be alternatives. This will improve code readability, making it easier to understand the function's logic.

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+        options?: GetPlaylistOptions,
+    ): Promise<PlaylistLibmuse> {
+        return retry<PlaylistLibmuse>(
+            async () => await get_playlist(playlistId, options),
```
It's recommended to use camelCase for variable naming in JavaScript

pkarpovich commented on src/services/music-providers/youtube-music/youtube-music.service.ts:
```diff
@@ -0,0 +1,286 @@
+        );
+    }
+
+    private async removePlaylistItemsWrapperOverLibmuseMethod(
```
I would suggest renaming `removePlaylistItemsWrapperOverLibmuseMethod`. Using a NPM package name in a function name is not a good practice as it can lead to confusion if the package is swapped or removed. A more generic, self-descriptive name like `removePlaylistItems` would improve readability and adhere to good coding practices.

pkarpovich commented on src/services/sync.service.ts:
```diff
@@ -106,7 +110,11 @@ export class SyncService {
 
-            await this.removeDeletedTracks(ctx, tracksForAdd);
+            if (
+                !target?.isNoRemoveTracksThatAreNotInOriginalPlaylistAndAreNoLongerAvailable
```
rename it

R00BEL committed:
refactoring: YoutubeMusicService moved to services import

R00BEL committed:
refactoring: artistsInDecomposed -> filteredArtists in YandexMusicService
refactoring: getPlaylistTracksLimit now equals Number.MAX_VALUE in YoutubeMusicService

R00BEL committed:
refactoring(YoutubeMusicService):
  - getting a function via property name
  - rename notifyThatFunctionCompletedWithError -> logApiError

R00BEL committed:
refactoring(YoutubeMusicService) rename:
  - checkThatTrackIsSuitable -> isTrackMatchingCriteria
  - findNameWithoutParticipantDesignations -> processedFindName
  - getTrackNameWithoutParticipantDesignations -> getPrimaryTrackName

R00BEL committed:
refactoring(YoutubeMusicService) rename:
  - add_playlist_items -> addPlaylistItems
  - get_option -> getOption
  - get_playlist -> getPlaylist
  - remove_playlist_items -> removePlaylistItems

R00BEL committed:
refactoring(YoutubeMusicService) rename:
  - remove_playlist_items -> removePlaylistItemsLibmuseMethode
  - removePlaylistItemsWrapperOverLibmuseMethod -> removePlaylistItemsWrapperOverApiMethod

pkarpovich commented:
Thanks a lot for this, and sorry for the long silence.

Adding YouTube Music turned out to be more trouble than it is worth: there is no official, stable API, so any integration has to lean on unofficial endpoints and cookie-based auth that break often and need ongoing maintenance. Given that cost, YouTube Music support is not something the project plans to take on right now.

On top of that, the codebase has since gone through a fairly large rewrite (the Yandex source, dependencies, and provider wiring all changed), so this branch is far out of date and would need to be redone from scratch against current main.

Closing for those reasons - I really appreciate the effort you put in here.
